File: Sorting/C++/quicksort.cpp
#include<iostream>

using namespace std;

void swap(int &a, int &b){
    int temp = a;
    a = b;
    b = temp;
}


int partition(int a[], int l, int r){
    int x=l, y=r;
    int pivot = a[x];

    while(x < y){
        while(a[x] <= pivot && x <= r){
            x++;
        }
        while(a[y] > pivot){
            y--;
        }

        if(x < y){
            swap(a[x], a[y]);
        }
    }

    swap(a[l], a[y]);
    return y;
}

void QuickSort(int a[], int l, int r){
    if(l<r){
        int p = partition(a, l, r);
        QuickSort(a, l, p-1);
        QuickSort(a, p+1, r);
    }
}

void printArray(int a[], int n){
    for(int i=0; i<n; i++){
        cout << a[i] << "\t";
    }
    cout << endl;
}

int main(){
    int a[] = {12, 17, 2, -1, 0, 1, 15, 100, 8, 19};

    int size = sizeof(a)/sizeof(a[0]);

    printArray(a, size);
    QuickSort(a, 0, size-1);
    printArray(a, size);
    
}

File: Sorting/Python/quicksort.py
def partition(arr, l, r):
    x, y = l, r
    pivot = arr[l]

    while x<y:
        while x<=r and arr[x] <= pivot:
            x+=1
        while arr[y] > pivot:
            y-=1
        
        if(x < y):
            arr[x], arr[y] = arr[y], arr[x]

    arr[l], arr[y] = arr[y], arr[l]  
    return y

def quick_sort(arr, l, r):
    if l<r:
        p = partition(arr, l, r);
        quick_sort(arr, l, p-1)
        quick_sort(arr, p+1, r)
        

arr = [12, 17, 2, -1, 0, 1, 15, 100, 8, 19]
print(arr)
quick_sort(arr, 0, 9)
print(arr)